example/train_onnx_model.py

```diff
@@ -0,0 +1,198 @@
+
+  onnx.save(onnx_model, trained_onnx_model)
+
+
+def train_tf_model():
+  import tensorflow as tf
+  from tensorflow.keras import datasets, layers, models
+  model = models.Sequential()
+  model.add(
+      layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
+  model.add(layers.MaxPooling2D((2, 2)))
+  model.add(layers.Conv2D(64, (3, 3), activation='relu'))
+  model.add(layers.MaxPooling2D((2, 2)))
+  model.add(layers.Conv2D(64, (3, 3), activation='relu'))
+
+  model.add(layers.Flatten())
+  model.add(layers.Dense(64, activation='relu'))
+  model.add(layers.Dense(10))
+
+  if not os.path.exists(saved_model_path):
+    os.mkdir(saved_model_path)
+  model.save(saved_model_path)
+  model.summary()
+  print("tf saved model: {}".format(saved_model_path))
+  '''
+    (train_images, train_labels), (test_images, test_labels) = datasets.cifar10.load_data()
+    train_images, test_images = train_images / 255.0, test_images / 255.0
+    
+    model.compile(optimizer='adam',
+              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+              metrics=['accuracy'])
+
+    history = model.fit(train_images, train_labels, epochs=4, 
+                    validation_data=(test_images, test_labels))
+    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
+    model.save(saved_model_path)
+    '''
+
+
+def convert_tf2onnx():
+  if not os.path.exists(onnx_model_path):
+    os.mkdir(onnx_model_path)
+  os.system('python -m tf2onnx.convert --saved-model {} --output {}'.format(
+      saved_model_path, onnx_model_file))
+  print('onnx model: {}'.format(onnx_model_file))
+
+
+def train_onnx_model():
+  onnx_model = onnx.load(onnx_model_file)
+  tf_rep = prepare(onnx_model, training_mode=True, logging_level=logging.ERROR)
+  training_flag_placeholder = tf_rep.tensor_dict[
+      onnx_tf.backend.training_flag_name]
+  input_name = onnx_model.graph.input[0].name
+  output_name = onnx_model.graph.output[0].name
+
+  with tf_rep.graph.as_default():
+    with tf_compat.Session() as sess:
+      y_truth = tf_compat.placeholder(tf.int64, [None], name='y-input')
+      tf_rep.tensor_dict["y_truth"] = y_truth
+      loss_op = tf.reduce_mean(
+          tf_compat.losses.sparse_softmax_cross_entropy(
+              labels=tf_rep.tensor_dict['y_truth'],
+              logits=tf_rep.tensor_dict[output_name]))
+      opt_op = tf_compat.train.AdamOptimizer().minimize(loss_op)
+      eval_op = tf.reduce_mean(input_tensor=tf.cast(
+          tf.equal(tf.argmax(input=tf_rep.tensor_dict[output_name], axis=1),
+                   tf_rep.tensor_dict['y_truth']), tf.float32))
+
+      train_data, test_data = get_dataset()
+      sess.run(tf_compat.global_variables_initializer())
+      print("==> Train the model..")
+
+      for epoch in range(1, epochs + 1):
+        step = 1
+        next_batch = tf.compat.v1.data.make_one_shot_iterator(
+            train_data).get_next()
```

Comment:
Got errors below, any idea how to fix it? Does it require minimum free memory?
2021-02-04 00:35:22.401127: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 1228800000 exceeds 10% of free system memory.
AttributeError: 'BatchDataset' object has no attribute '_make_one_shot_iterator'

Reply:
@chinhuang007 May I know your tensorflow version?

Reply:
@chinhuang007 I did not hit this issue with tf2.4 and tf2.3.1 in both mac and linux environment. What's your test environment?

Reply:
Python version:
3.6.9 (default, Nov  7 2019, 10:44:02)
[GCC 8.3.0]
ONNX version:
1.8.0
ONNX-TF version:
1.7.0
Tensorflow version:
2.4.0

Reply:
I updated cifar10 dataset to mnist, the memory requirement should be reduced from 15G to 5G.